src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -21,149 +28,148 @@ public sealed override string ConvertName(string name)
                 }
-
-                resultUsedLength += written;
-            }
-
-            int first = 0;
-            ReadOnlySpan<char> chars = name.AsSpan();
-            CharCategory previousCategory = CharCategory.Boundary;
-
-            for (int index = 0; index < chars.Length; index++)
-            {
-                char current = chars[index];
-                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
-
-                if (currentCategoryUnicode == UnicodeCategory.SpaceSeparator ||
-                    currentCategoryUnicode >= UnicodeCategory.ConnectorPunctuation &&
-                    currentCategoryUnicode <= UnicodeCategory.OtherPunctuation)
+                else if (current == separator)
                 {
-                    WriteWord(chars.Slice(first, index - first), ref result);
-
-                    previousCategory = CharCategory.Boundary;
-                    first = index + 1;
-
-                    continue;
+                    // Json.NET compat: reset state if the separator character is encountered in the input string.
+                    WriteChar(separator, ref destination);
+                    state = SeparatorState.NotStarted;
                 }
-
-                if (index + 1 < chars.Length)
+                else
                 {
-                    char next = chars[index + 1];
-                    CharCategory currentCategory = currentCategoryUnicode switch
-                    {
-                        UnicodeCategory.LowercaseLetter => CharCategory.Lowercase,
-                        UnicodeCategory.UppercaseLetter => CharCategory.Uppercase,
-                        _ => previousCategory
-                    };
-
-                    if (currentCategory == CharCategory.Lowercase && char.IsUpper(next) ||
-                        next == '_')
+                    // Handle all remaining character categories
+                    if (state is SeparatorState.SpaceSeparator)
                     {
-                        WriteWord(chars.Slice(first, index - first + 1), ref result);
-
-                        previousCategory = CharCategory.Boundary;
-                        first = index + 1;
-
-                        continue;
+                        // Normalize preceding spaces to one separator.
+                        WriteChar(separator, ref destination);
                     }
 
-                    if (previousCategory == CharCategory.Uppercase &&
-                        currentCategoryUnicode == UnicodeCategory.UppercaseLetter &&
-                        char.IsLower(next))
-                    {
-                        WriteWord(chars.Slice(first, index - first), ref result);
-
-                        previousCategory = CharCategory.Boundary;
-                        first = index;
-
-                        continue;
-                    }
+                    if (!lowercase)
+                        current = char.ToUpperInvariant(current);
```

Comment:
Nit: Let's keep bracing consistent in the same method. Just above this there's a one-line body that uses braces, it'd be nice to do so here, too.